lib/seattleflu/db/cli/command/etl/presence_absence.py

```diff
@@ -0,0 +1,388 @@
+
+                        # Most of the time we expect to see new samples and new
+                        # presence_absence tests, so an insert-first approach makes more sense.
+                        # Presence-absence tests we see more than once are presumed to be
+                        # corrections.
+                        upsert_presence_absence(db,
+                            identifier = test_result["id"],
+                            sample_id  = sample.id,
+                            target_id  = target.id,
+                            present    = get_target_result(test_result["targetStatus"], 
+                                                           sample.identifier, target.identifier),
+                            well_details = test_result["wellResults"])
+
+                mark_processed(db, group.id)
+
+                LOG.info(f"Finished processing presence_absence group {group.id}")
+
+    except Exception as error:
+        processed_without_error = False
+
+        LOG.error(f"Aborting with error")
+        raise error from None
+
+    else:
+        processed_without_error = True
+
+    finally:
+        if action == "prompt":
+            ask_to_commit = \
+                "Commit all changes?" if processed_without_error else \
+                "Commit successfully processed enrollments up to this point?"
+
+            commit = click.confirm(ask_to_commit)
+        else:
+            commit = action == "commit"
+
+        if commit:
+            LOG.info(
+                "Committing all changes" if processed_without_error else \
+                "Committing successfully processed enrollments up to this point")
+            db.commit()
+
+        else:
+            LOG.info("Rolling back all changes; the database will not be modified")
+            db.rollback()
+
+def find_or_create_target(db: DatabaseSession, identifier: str, control_status: bool) -> Any:
+    """
+    Select presence_absence test target by *identifier*, or insert it if it doesn't exist.
+    """
+    LOG.debug(f"Looking up target «{identifier}»")
+
+    target = db.fetch_row("""
+        select target_id as id, identifier
+          from warehouse.target
+         where identifier = %s
+        """, (identifier,))
+
+    if target:
+        LOG.info(f"Found target {target.id} «{target.identifier}»")
+    else:
+        LOG.debug(f"Target «{identifier}» not found, adding")
+
+        data = {
+            "identifier": identifier,
+            "control_status": control_status
+        }
+
+        target = db.fetch_row("""
+            insert into warehouse.target (identifier, control_status)
+                values (%(identifier)s, %(control_status)s)
+            returning target_id as id, identifier
+            """, data)
+
+        LOG.info(f"Created target {target.id} «{target.identifier}»")
+
+    return target
+
+
+def target_control_status(control_status: str, target_identifier: str) -> bool:
```

Comment:
I think it might be cleaner not to pass in arguments to a function just for logging purposes, e.g. arguments which don't affect the function's behavior or return value. What do you think about dropping the `for target {target_identifier}` bit from the log message (the identifier will be in immediately preceding log lines) so that we can drop the function argument?

Reply:
Going to add additional logging statements to the top of each for loop and then remove extra logging-only arguments from value transformation functions.